### PR TITLE
Feature/django-fixes

### PR DESCRIPTION
**Pull request by gecBurton**
## Context

Various fixes for the django app exposed during deployment:

* enable local running/testing/debugging
  * optionally turn off compression (MacOS can use linux SASS binaries)
  * add django settings to `pyproject.toml`
* serve admin static files
  * restore `AppDirectoriesFinder`
* auth fixes
  * make password optional so that user can be added without one
  * remove password specific code
  * remove non existent auth backend
* fix integration tests

## Things to check

- [ ] I have added any new ENV vars in all deployed environments
- [x] I have tested any code added or changed
- [x] I have run integration tests https://github.com/i-dot-ai/redbox-copilot/actions/runs/9059162137

### CODE REVIEW

File: django_app/redbox_app/redbox_core/models.py [hunk 1]
```diff
@@ -3,11 +3,8 @@
 from django.conf import settings
 from django.db import models
 from django_use_email_as_username.models import BaseUser, BaseUserManager
-from dotenv import load_dotenv
 from jose import jwt
 
-load_dotenv()
```
Comment (gecBurton) [inline django_app/redbox_app/redbox_core/models.py:9]: this belongs in `settings.py`

File: django_app/redbox_app/redbox_core/models.py [hunk 2]
```diff
@@ -32,6 +29,7 @@ class User(BaseUser, UUIDPrimaryKeyBase):
     invited_at = models.DateTimeField(default=None, blank=True, null=True)
     invite_accepted_at = models.DateTimeField(default=None, blank=True, null=True)
     last_token_sent_at = models.DateTimeField(editable=False, blank=True, null=True)
+    password = models.CharField("password", max_length=128, blank=True, null=True)
```
Comment (gecBurton) [inline django_app/redbox_app/redbox_core/models.py:32]: the only override here is `null=True` which ensures that the admin doesnt ask us for a password when creating a new user

File: django_app/redbox_app/settings.py [hunk 3]
```diff
@@ -19,21 +22,24 @@
...
 BASE_DIR = Path(__file__).resolve().parent.parent
 
-COMPRESS_ENABLED = True
+COMPRESSION_ENABLED = env.bool("COMPRESSION_ENABLED")
+
 COMPRESS_PRECOMPILERS = (("text/x-scss", "django_libsass.SassCompiler"),)
 
 STATIC_URL = "static/"
-STATIC_ROOT = "frontend/"
+STATIC_ROOT = "staticfiles/"
```
Comment (gecBurton) [inline django_app/redbox_app/settings.py:30]: this appears to be the standard way of doing this: multiple directories to keep your static files but just one (different) `STATIC_ROOT` where they all get copied to on running `python manage.py collectstatic`

File: django_app/redbox_app/settings.py [hunk 4]
```diff
@@ -19,21 +22,24 @@
...
+    os.path.join(BASE_DIR, "frontend/"),
+    # (
+    #     "govuk-assets",
+    #     BASE_DIR / "frontend/node_modules/govuk-frontend/dist/govuk/assets",
+    # ),
 ]
 STATICFILES_FINDERS = [
     "compressor.finders.CompressorFinder",
     "django.contrib.staticfiles.finders.FileSystemFinder",
+    "django.contrib.staticfiles.finders.AppDirectoriesFinder",
```
Comment (gecBurton) [inline django_app/redbox_app/settings.py:42]: this is needed to collect the admin static files

File: django_app/redbox_app/settings.py [hunk 5]
```diff
@@ -123,15 +133,6 @@
     {
         "NAME": "django.contrib.auth.password_validation.NumericPasswordValidator",
     },
-    {
-        "NAME": "redbox_app.custom_password_validators.SpecialCharacterValidator",
```
Comment (gecBurton) [inline django_app/redbox_app/settings.py:127]: we are using passwords

File: django_app/redbox_app/templates/base.html [hunk 6]
```diff
@@ -14,9 +14,11 @@
   <link rel="manifest" href="{{static("govuk-assets/manifest.json")}}">
 
   <meta name="robots" content="noindex, nofollow">
-  {% compress css %}
-  <link rel="stylesheet" type="text/x-scss" href="{{static('style.scss')}}" />
-  {% endcompress %}
+  {% if COMPRESSION_ENABLED %}
```
Comment (gecBurton) [inline django_app/redbox_app/templates/base.html:17]: this allows us to turn compression off

File: django_app/redbox_app/settings.py [hunk 7]
```diff
@@ -82,11 +88,16 @@
             BASE_DIR / "redbox_app" / "templates",
             BASE_DIR / "redbox_app" / "templates" / "auth",
         ],
-        "OPTIONS": {"environment": "redbox_app.jinja2.environment"},
+        "OPTIONS": {
+            "environment": "redbox_app.jinja2.environment",
+            "context_processors": [
+                "redbox_app.context_processors.compression_enabled",
```
Comment (gecBurton) [inline django_app/redbox_app/settings.py:94]: this allows us to turn compression off

File: django_app/redbox_app/urls.py [hunk 8]
```diff
@@ -38,4 +40,9 @@
     path("health/", views.health, name="health"),
 ]
 
-urlpatterns = info_urlpatterns + other_urlpatterns + auth_urlpatterns
+urlpatterns = (
+    info_urlpatterns
+    + other_urlpatterns
+    + auth_urlpatterns
+    + static(settings.STATIC_URL, document_root=settings.STATIC_ROOT)
```
Comment (gecBurton) [inline django_app/redbox_app/urls.py:47]: we should remove this

File: django_app/tests/conftest.py [hunk 9]
```diff
@@ -13,6 +14,11 @@
 logger = logging.getLogger(__name__)
 
 
+@pytest.fixture(autouse=True, scope="session")
+def collect_static():
```
Comment (gecBurton) [inline django_app/tests/conftest.py:18]: this is needed for local testing

File: django_app/redbox_app/settings.py [hunk 10]
```diff
@@ -104,7 +115,6 @@
 
 AUTHENTICATION_BACKENDS = [
     "django.contrib.auth.backends.ModelBackend",
-    "magic_link.backends.MagicLinkBackend",
```
Comment (gecBurton) [inline django_app/redbox_app/settings.py:107]: this doesnt exist (maybe from another implementation https://pypi.org/project/django-magiclink/ that we aren using?)

File: redbox/models/settings.py [hunk 11]
```diff
@@ -65,7 +65,7 @@ class Settings(BaseSettings):
     django_settings_module: str = "redbox_app.settings"
     debug: bool = True
     django_secret_key: str
-    django_log_level: str
+    django_log_level: Literal["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"] = "WARNING"
```
Comment (gecBurton) [inline redbox/models/settings.py:68]: https://docs.djangoproject.com/en/5.0/topics/logging/#loggers

File: django_app/redbox_app/settings.py (hunk 3 above)
Comment (brunns) [reply django_app/redbox_app/settings.py:30]: Do we need to add `python manage.py collectstatic` to the `Makefile` somewhere?

File: redbox/models/settings.py (hunk 11 above)
Comment (brunns) [reply redbox/models/settings.py:68]: We should probably just remove the setting altogether, and make extras optional as per https://stackoverflow.com/a/77021677/1073696.

**Review by brunns — COMMENTED** [review]
Couple of nitpicks.

File: .env.test [hunk 12]
```diff
@@ -66,3 +66,5 @@ FROM_EMAIL=test@example.com
 GOVUK_NOTIFY_PLAIN_EMAIL_TEMPLATE_ID=example-id
 EMAIL_FILE_PATH='/app/mail'
 USE_STREAMING=False
+COMPRESSION_ENABLED=False
+AWS_REGION=eu-west-
```
Comment (252afh) [inline .env.test:70]: missing `2` on the end of region var

File: django_app/redbox_app/redbox_core/migrations/0004_alter_user_password.py [hunk 13]
```diff
@@ -0,0 +1,19 @@
+# Generated by Django 5.0.6 on 2024-05-12 18:36
+
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("redbox_core", "0003_chathistory_file_chatmessage_and_more"),
```
Comment (252afh) [inline django_app/redbox_app/redbox_core/migrations/0004_alter_user_password.py:8]: I think the dependency here will need to be stepped now

File: django_app/redbox_app/settings.py (hunk 3 above)
Comment (gecBurton) [reply django_app/redbox_app/settings.py:31]: good idea

**Review by 252afh — APPROVED** [review]

File: django_app/redbox_app/redbox_core/migrations/0004_alter_user_password.py (hunk 13 above)
Comment (gecBurton) [reply django_app/redbox_app/redbox_core/migrations/0004_alter_user_password.py]: https://github.com/i-dot-ai/redbox-copilot/pull/341/commits/521f10e1d9870aef92d321bf17cdd374f1d05814
good catch

File: .env.test (hunk 12 above)
Comment (gecBurton) [reply .env.test]: [7679a9e](https://github.com/i-dot-ai/redbox-copilot/pull/341/commits/7679a9ea62cd0e0bfba4f5599c2fd5fc4e0107d2)

File: django_app/redbox_app/redbox_core/migrations/0004_alter_user_password.py (hunk 13 above)
Comment (gecBurton) [reply django_app/redbox_app/redbox_core/migrations/0004_alter_user_password.py]: [521f10e](https://github.com/i-dot-ai/redbox-copilot/pull/341/commits/521f10e1d9870aef92d321bf17cdd374f1d05814)

File: redbox/models/settings.py (hunk 11 above)
Comment (gecBurton) [reply redbox/models/settings.py:65]: [0d7a495](https://github.com/i-dot-ai/redbox-copilot/pull/341/commits/0d7a4950c92b7c4db9ac52e2d7354d31f39e14c7)

**Pull request merged**